File: core/src/app/add_wall_handler.cpp
#include "toporoom/app/add_wall_handler.hpp"

#include "toporoom/domain/floor_plan_document.hpp"
#include "toporoom/domain/length_mm.hpp"
#include "toporoom/domain/point_mm.hpp"

namespace toporoom::app {

CommandResult AddWallHandler::execute(const AddWallCommand& command) {
  auto document = load_document(store_, command.document_id);
  domain::AddWallProps props;
  props.storey_id = command.storey_id;
  props.id = command.wall_id;
  props.start = domain::PointMm::of(command.start_x, command.start_y);
  props.end = domain::PointMm::of(command.end_x, command.end_y);
  props.thickness = domain::LengthMm::of(command.thickness_mm);
  props.height = domain::LengthMm::of(command.height_mm);
  props.kind = command.kind;
  document.add_wall(std::move(props));
  return commit(store_, rebuild_, document);
}

}  // namespace toporoom::app

File: packages/app-floorplan/src/add-wall-handler.ts
import { LengthMm, PointMm } from "@toporoom/domain-floorplan";
import type { WallKind } from "@toporoom/domain-floorplan";
import type { DocumentStorePort } from "@toporoom/ports";
import { commit, loadDocument, type CommandResult } from "./document-io.js";
import type { GeometryRebuildPolicy } from "./rebuild-policy.js";

export interface AddWallCommand {
  documentId: string;
  storeyId: string;
  wallId?: string;
  start: { x: number; y: number };
  end: { x: number; y: number };
  thicknessMm: number;
  heightMm: number;
  kind: WallKind;
}

export class AddWallHandler {
  constructor(
    private readonly store: DocumentStorePort,
    private readonly rebuild: GeometryRebuildPolicy,
  ) {}

  async execute(command: AddWallCommand): Promise<CommandResult> {
    const document = await loadDocument(this.store, command.documentId);
    document.addWall({
      storeyId: command.storeyId,
      id: command.wallId,
      start: PointMm.of(command.start.x, command.start.y),
      end: PointMm.of(command.end.x, command.end.y),
      thickness: LengthMm.of(command.thicknessMm),
      height: LengthMm.of(command.heightMm),
      kind: command.kind,
    });
    return commit(this.store, this.rebuild, document);
  }
}
